### PR TITLE
Recheck the failure or completion status of missions after any mission is offered, completed, or failed on a planet.

### PR DESCRIPTION
**Pull request by NomadicVolcano**
This resolves a long-standing bug: mission failure and completion status is not updated after other missions are processed.  See #1938, #4759, and #4761.  There are probably other issues caused by this bug.

This only fixes the bug at planet-side.  No extra logic is added to correct mission failures and completions during combat.

The possible side-effect of this fix is that some missions that were relying on this bug may complete or fail a day or two earlier, when their completion or failure conditions are met.

I have a test mission sequence for this bug.

[test.txt](https://github.com/endless-sky/endless-sky/files/4113639/test.txt)

To use it:

1. Visit Follower
2. You should now see missions A0 through A4 on the job board.  Destinations will be Valhalla or Earth
3. Visit the spaceport at Follower
4. You should now see missions B1 through B4 with a destination of Earth
5. Fly to Valhalla
6. Missions A0 through A4 will send messages.
7. Mission A0 will be done and A1 through A4 will be failed in your saved game
8. Visit the spaceport at Valhalla
9. Accept mission B0
10. Missions B1 through B4 will send messages.
11. Missions B0 through B4 will be failed in your saved game

### CODE REVIEW

**Comment by NomadicVolcano** [comment]
I don't like leaving ancient pull requests and ancient branches. I haven't even seen a review yet. Is this fix actually wanted? If not, I'll delete the branch and close the PR.

**Comment by NomadicVolcano** [comment]
I'm closing the PR. Someone can re-open it if they want. I'll wait another week or two to delete the branch.

**Comment by Amazinite** [comment]
> Someone can re-open it if they want.

Okay.

**Comment by NomadicVolcano** [comment]
> > Someone can re-open it if they want.
> 
> Okay.

You might want to grab the branch too. I'm going through my list of branches deleting any that are unneeded. I've made an annotated tag for it, but I'm not keeping a disconnected branch around for much longer.

**Comment by Amazinite** [comment]
> I don't like leaving ancient pull requests and ancient branches. 

Two months is hardly ancient. Most of your PRs will end up closed at that rate. No need to be impatient.

**Comment by NomadicVolcano** [comment]
> > Someone can re-open it if they want.
> 
> Okay.

I suppose I could consider your response some activity on the issue. I'll keep it open a little longer and respond to reviews. If I don't see even a request for review by mid-April, then I'll give up on it with finality.

**Comment by NomadicVolcano** [comment]
> > I don't like leaving ancient pull requests and ancient branches.
> 
> Two months is hardly ancient. Most of your PRs will end up closed at that rate. No need to be impatient.

There wasn't a review, or even a comment. To me, that means the PR is wanted by nobody but me.

**Comment by Amazinite** [comment]
I guess the milestones mean nothing then.

**Comment by NomadicVolcano** [comment]
> I guess the milestones mean nothing then.

I did not notice the milestone change.

If you do merge this, it should be as early as possible in a release cycle, preferably an unstable release (0.9.13 or 0.9.15). It shouldn't break anything, but I can't promise that. There might be a mission somewhere that is relying on bug #1938. The entire game should be retested.

**Comment by NomadicVolcano** [comment]
I will try to explain this another way. The amount of testing for this commit will be larger than all of my other PRs combined. Even after I PR the Korath Exile changes, which require rebalancing two dozen systems, it probably will still require more testing than all my other PRs. I don't want a test that big hanging over me forever.

**Comment by Amazinite** [comment]
> There might be a mission somewhere that is relying on bug #1938. The entire game should be retested.
>
> I don't want a test that big hanging over me forever.

I have no issue starting a new pilot and playing through the entire game to see if anything is broken. Although, I can't recall any missions that rely on the existing behavior.

**Comment by NomadicVolcano** [comment]
> > There might be a mission somewhere that is relying on bug #1938. The entire game should be retested.
> > I don't want a test that big hanging over me forever.
> 
> I have no issue starting a new pilot and playing through the entire game to see if anything is broken. Although, I can't recall any missions that rely on the existing behavior.

I played through all the major stories with this bug fix, but not the smaller ones. I also tried some of my story PRs that had bugs, and the bugs vanished.

**Comment by NomadicVolcano** [comment]
I was told this PR has not been reviewed since it is not needed by any stories. I had to rewrite about 400 lines of missions in MasterOfGrey's #4952 because his early dialog missions didn't work with bug #1938. Also, story #4901 *can't* work around bug #1938, so I had to rewrite parts of the story to get it working with the bug.

I am certain the reason this PR has not been reviewed is that I asked for it to be a priority, back in January. The community is suffering as a result, and I don't think this PR will be reviewed unless someone else submits a fix.

I'll wait another week or so before wiping the branch and closing the PR. If someone else wants to take it over, or PR an alternative fix, please let me know.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -2427,19 +2452,30 @@ void PlayerInfo::StepMissions(UI *ui)
...
-	// more time to see if any mission is now completed or failed due to a change
-	// that happened in another mission the first time through.
-	mit = missions.begin();
-	while(mit != missions.end())
+	// One mission's actions may influence another mission, so loop through
+	// to see if any mission is now completed or failed due to a change
+	// that happened in earlier passes.  For safety's sake, we limit the
+	// number of passes to 30.
+	bool modifiedMissions=true;
+	for(int i=0;i<30 && modifiedMissions;i++)
```
[the remark this answers is not in the document]
Comment (NomadicVolcano) [reply source/PlayerInfo.cpp]: This loop is no longer needed. I deleted it in the latest version.

File: source/PlayerInfo.cpp [hunk 2]
```diff
@@ -1501,11 +1501,11 @@ void PlayerInfo::ClearActiveBoardingMission()
 // If one of your missions cannot be offered because you do not have enough
 // space for it, and it specifies a message to be shown in that situation,
 // show that message.
-void PlayerInfo::HandleBlockedMissions(Mission::Location location, UI *ui)
+bool PlayerInfo::HandleBlockedMissions(Mission::Location location, UI *ui)
```
[the remark this answers is not in the document]
Comment (NomadicVolcano) [reply source/PlayerInfo.cpp]: Correct. I removed it in the latest version.

File: source/PlanetPanel.cpp [hunk 3]
```diff
@@ -233,6 +235,10 @@ void PlanetPanel::TakeOffIfReady()
 		return;
 	}
 	
+	// Update the completion and failure status of missions:
+	if(player.RecheckMissions(GetUI()))
```
[the remark this answers is not in the document]
Comment (NomadicVolcano) [reply source/PlanetPanel.cpp:239]: This line is the crux of the bug fix. Once per game step (ie. after each conversation dialog) the completion and failure status is checked. Hence, when a mission is accepted, failed, or whatever, all missions are tested again.

The other loop you saw before does nothing; I've deleted it.

**Comment by NomadicVolcano** [comment]
> This feels like a hack of a solution, rather than the proper solution.

The only hacky aspect was the old loop I forgot to remove.

The actual bug fix is the RecheckMissions function, and the behavior of checking mission failure and completion conditions after 

> a large refactoring

No matter what refactoring you do, you'll still need to recheck failure and completion conditions multiple times. Otherwise mission dependencies won't work, and you'll get this same bug all over again in elegant, newly-refactored, code.

**Comment by NomadicVolcano** [comment]
Closing all pull requests related to this account.

**Pull request closed without merging**